### ISSUE TITLE
Refactor TableView to use asyncinject

### ISSUE THREAD

**Issue by simonw** [simonw/datasette#1715]
I've been working on a dependency injection mechanism in a separate library:

- https://github.com/simonw/asyncinject

I think it's ready to try out with Datasette to see if it's a pattern that will work here.

I'm going to attempt to refactor `TableView` to use it. There are two overall goals here:

- Use `asyncinject` to add parallel execution of some aspects of the table page - most notably I want to be able to execute the `count(*)` query, the `select ...` query, the various faceting queries and the facet suggestion queries in parallel - and measure if doing so is good for performance.
- Use it to execute different output formats (possibly with some changes to the existing `register_output_renderer()` plugin hook). I want CSV and JSON to use the same mechanism that plugins use.

Stretch goal is to get this working with streaming data too, see:

- #1101

**Comment by simonw** [comment]
Looking at the start of `TableView.data()`:

https://github.com/simonw/datasette/blob/d57c347f35bcd8cff15f913da851b4b8eb030867/datasette/views/table.py#L333-L346

I'm going to resolve `table_name` and `database` from the URL - `table_name` will be a string, `database` will be the DB object returned by `datasette.get_database()`. Then those can be passed in separately too.

**Comment by simonw** [comment]
```python
async def database(request: Request, datasette: Datasette) -> Database:
    database_route = tilde_decode(request.url_vars["database"])
    try:
        return datasette.get_database(route=database_route)
    except KeyError:
        raise NotFound("Database not found: {}".format(database_route))

async def table_name(request: Request):
    return tilde_decode(request.url_vars["table"])
```

**Comment by simonw** [comment]
I'm having second thoughts about injecting `request` - might be better to have the view function pull the relevant pieces out of the request before triggering the rest of the resolution.

**Comment by simonw** [comment]
The `RowTableShared` class is making this a whole lot more complicated.

I'm going to split the `Row` view out into an entirely separate `views/row.py` module.

**Comment by simonw** [comment]
Pushed my WIP on this to the `api-extras` branch: 5053f1ea83194ecb0a5693ad5dada5b25bf0f7e6

**Comment by simonw** [comment]
The refactor I did in #1719 pretty much clashes with all of the changes in https://github.com/simonw/datasette/commit/5053f1ea83194ecb0a5693ad5dada5b25bf0f7e6 so I'll probably need to start my `api-extras` branch again from scratch.

**Comment by simonw** [comment]
This time I'm not going to bother with the `filter_args` thing - I'm going to just try to use `asyncinject` to execute some big high level things - facets, suggested facets, counts, the query - and then combine it with the `extras` mechanism I'm trying to introduce too.

**Comment by simonw** [comment]
I'm going to rename `database` to `database_name` and `table` to `table_name` to avoid confusion with the `Database` object as opposed to the string name for the database.

**Comment by simonw** [comment]
`pytest tests/test_table_*` runs the tests quickly.

**Comment by simonw** [comment]
Well this is fun... I applied this change:

```diff
diff --git a/datasette/views/table.py b/datasette/views/table.py
index d66adb8..85f9e44 100644
--- a/datasette/views/table.py
+++ b/datasette/views/table.py
@@ -1,3 +1,4 @@
+import asyncio
 import itertools
 import json
 
@@ -5,6 +6,7 @@ import markupsafe
 
 from datasette.plugins import pm
 from datasette.database import QueryInterrupted
+from datasette import tracer
 from datasette.utils import (
     await_me_maybe,
     CustomRow,
@@ -174,8 +176,11 @@ class TableView(DataView):
                 write=bool(canned_query.get("write")),
             )
 
-        is_view = bool(await db.get_view_definition(table_name))
-        table_exists = bool(await db.table_exists(table_name))
+        with tracer.trace_child_tasks():
+            is_view, table_exists = map(bool, await asyncio.gather(
+                db.get_view_definition(table_name),
+                db.table_exists(table_name)
+            ))
 
         # If table or view not found, return 404
         if not is_view and not table_exists:
```
And now using https://datasette.io/plugins/datasette-pretty-traces I get this:

![CleanShot 2022-04-26 at 14 03 33@2x](https://user-images.githubusercontent.com/9599/165392009-84c4399d-3e94-46d4-ba7b-a64a116cac5c.png)

**Comment by simonw** [comment]
Running facets and facet suggestions in parallel using `asyncio.gather()` turns out to be a lot less hassle than I had thought - maybe I don't need `asyncinject` for this at all?

```diff
         if not nofacet:
-            for facet in facet_instances:
-                (
-                    instance_facet_results,
-                    instance_facets_timed_out,
-                ) = await facet.facet_results()
+            # Run them in parallel
+            facet_awaitables = [facet.facet_results() for facet in facet_instances]
+            facet_awaitable_results = await asyncio.gather(*facet_awaitables)
+            for (
+                instance_facet_results,
+                instance_facets_timed_out,
+            ) in facet_awaitable_results:
                 for facet_info in instance_facet_results:
                     base_key = facet_info["name"]
                     key = base_key
@@ -522,8 +540,10 @@ class TableView(DataView):
             and not nofacet
             and not nosuggest
         ):
-            for facet in facet_instances:
-                suggested_facets.extend(await facet.suggest())
+            # Run them in parallel
+            facet_suggest_awaitables = [facet.suggest() for facet in facet_instances]
+            for suggest_result in await asyncio.gather(*facet_suggest_awaitables):
+                suggested_facets.extend(suggest_result)
```

**Comment by simonw** [comment]
I'm not going to use `asyncinject` in this refactor - at least not until I really need it. My research in these issues has put me off the idea ( in favour of `asyncio.gather()` or even not trying for parallel execution at all):

- #1727
